### ISSUE TITLE
Migration of tokens and identities from one institution to another institution

### ISSUE THREAD

**Issue by phavekes** [OpenConext/Stepup-Project#350]
[This issue is imported from pivotal - Originaly created at Apr 19, 2021 by Pieter van der Meulen](https://www.pivotaltracker.com/story/show/177825802)

	The immediate reason for this story is that we have two institutions A and B that are merging and that together will form a new institution C. Many of their users now have a Stepup Identity in institution A and B with active token. We want to find a solution whereby we can do the migration without the user have to re-register their token. Preferably the user has to do nothing extra and "magically" and up with a new identity with their existing tokens already vetted.

So we want to create new identities, with vetted tokens, in C based on the current active tokens of the user. To this end the institution will provide us with a list of the old identities in A and B to migrate to new identities in C. So for all identities we will know: 

<old NameID or uid> <old institution> <new NameID or uid> <new email>

For users that now have Identities in A and B, the institution will decide which identity to migrate. All the three institutions will be active for some time in Stepup while the institution migrates their applications to use institution C. After that process is finished the institutions A and B are disabled (and the users can be deprovisioned). Note that the uid and email address of the user in the new institution C may be different from the their old uid and email. Their name would typically be the same but could also be different due to spelling and formatting. 

In general migrating – or more precise creating a new one based on an existing one – an Identity together with the active tokens associated to a new institution should be possible. (And I can also see usecases where copying identities within the institution is usefull, for instance when an institution switches to a new IDM and uid\'s change)

The audit log must reflect that a token was "vetted" because it was copied from an identity in another institution.

Apart from the vetting requirement all the other checks done by the middleware during a migration should still apply. E.g. (not a complete list!):
- Only enabled token types
- No more than one token per (GSSP) type

*Implementation*
To implement this we are thinking about slightly extending the existing middleware console interface so it can be used for creating the new identities and tokens. The command are then prepared beforehand and run in batches. The console interface currently allows creating identities and tokens for bootstrapping a new Stepup installation and for adding tokens to existing identities for debugging purposes.

The existing console command would already allow us to do this. We can create a new identity with vetted tokens based on the second_factor_identifier (i.e. phone number, yubikey id, tiqr gssp id). But because the interface was not created with this usecase in mind, some features are missing.

- When an error occurs this must be reflected in the return code. I.e. return 0 on success, 1 on failure. Now the console returns 0.
- Ensure that each console command is transactional. There must be no effect when an error occurs. Individual events are already transaction, but when multiple event are involved this may not be the case. E.g. YubikeySecondFactorBootstrappedEvent is not transactional. We need to verify this for the commands that we will be using.
- The middleware:bootstrap:[gssp,sms,yubikey] commands (and their associated events) require an actor-id of an RA. We could use the identity-id of an SRAA for this. Better would be a way to store what happend (i.e. that this is a migrated token and identity) and show this in the audit log. That would mean changing the existing events, or creating new events.
- Idempotency or some other way (e.g. a return code/logging that indicate) to see that a command fails because the identity or token already exists. This could happen because
  - The user\'s identity already exists (the user did already login to the self service portal)
  - The batch is rerun because something failed the previous time.

*Event model*
Identities are created with:
`Surfnet.Stepup.Identity.Event.IdentityCreatedEvent`
The bootstrap console command has its own event (`Surfnet.Stepup.Identity.Event.YubikeySecondFactorBootstrappedEvent`)
The token creation events use a different event per TokenType (Yubikey, Phone, Gssf):
`Surfnet.Stepup.Identity.Event.[TokenType]PossessionProvenEvent`
`Surfnet.Stepup.Identity.Event.EmailVerifiedEvent`
`Surfnet.Stepup.Identity.Event.[TokenType]PossessionProvenAndVerifiedEvent`
And have a explicit vetting step:
`Surfnet.Stepup.Identity.Event.SecondFactorVettedEvent`

All these events are part of one eventstream that corresponds with the identity. Creating the new identity starts a new eventstream.
From there the projections are updated. These use identity UUID or token UUID as primary key, depending on wether they contain identities or tokens.

Would adding a Surfnet.Stepup.Identity.Event.ActivatedSecondFactorMigratedEvent be a good to implement this? This event can record the original token AND can verify that the token is currently active. Because a batch is created beforehand the user\'s token could have been revoked in the mean time.

A migration of a user would then consist of:
1. Surfnet.Stepup.Identity.Event.IdentityCreatedEvent
2. Surfnet.Stepup.Identity.Event.ActivatedSecondFactorMigratedEvent

This last "ActivatedSecondFactorMigratedEvent" event would:
- Require that the token to be migrated is currently active
- Update the projections. At least (check if more are needed):
  - middleware.vetted_second_factor
  - middleware.audit_log
  - gateway.second_factor

We could also introduce a similar event for the Identity ("IdentityMigratedEvent") to record in the event stream which identity was seen as the origin of the new identity. We could show this in the audit log as well. I do not see a strong usecase for this event though.

Let\'s discuss

**Comment by phavekes** [comment]
Nice work. This seems a good approach for this problem.
I think a ```Surfnet.Stepup.Identity.Event.IdentityMigratedEvent``` (or adding a \'migrated\'-flag to ```Surfnet.Stepup.Identity.Event.IdentityCreatedEvent```) would benefit the audit log.

Should identities with no active tokens be migrated, or should we only create/migrate identities with an active token? (Peter Havekes - Apr 20, 2021)

**Comment by phavekes** [comment]
@phavekes Thanks!

My idea is that migrating an identity without any active token would fail and that we migrate active tokens only. Migrating tokens awaiting email validation or activation would not help much and could even be unhelpful. If we copy them, then new email verification links and/or activation codes would be required, otherwise we would need to revoke the originals. Lots of bookkeeping and testing for us, and we don\'t even know if we helped or confused the user. The user probably needs an active token in the old institution as well. Revoked tokens are dead and I see no reason for migrating them. 

We could technically create the new identity, but this is not necessary for the migration. The user will have to login to the selfservice portal anyway to get a token. And when the user logs in to the selfservice portal the identity will be created anyway. (Pieter van der Meulen - Apr 20, 2021)

**Comment by phavekes** [comment]
As discussed:
- Use a new ActivatedSecondFactorMigratedEvent or add a new vetting type to the SecondFactorVettedEvent like we did for the vetting a token with another token? We\'re leaning towards a new event
    - Token migration has no existing identity as actor. We do not want to introduce a "system" identity, because that identity would be an exception adding complexity. Having a ActivatedSecondFactorMigratedEvent makes this clear.
    - What are effects on code complexity? Pro of ActivatedSecondFactorMigratedEvent: can add extra info (original token and identityid).

- No seperate IdentityMigratedEvent, it does not really add something. The ActivatedSecondFactorMigratedEvent can record the original token UUID, this implicitly leads to the Identity UUID. Explicitly adding the Identity UUID: better because it make this relation more accessible, no digging in database?

- What is the current Stepup-SelfService behaviour when an existing identity (nameid + institution) authenticates but the email address and/or commonName are different. At what point(s) / when are these updated?

- Console commands run from PHP CLI. The overhead of starting this is much higher than a REST call to PHP-FPM. Makes sense therefore to create one PHP CLI console command to process a batch. 
    - Include an option to specify a delay between identity migrations to control the load on the system. 
    - We want idempotency so we can rerun a batch in case of an error
    - Input would be CSV file with old and new identity information

- What is the behaviour when an Identity exists, but with a different commonName of email? Simplest is to make this an error.

- Cleaning up (deprovisioning) or revoking access to the old identities is not part of this story. There is existing functionality in Stepup for this (whitelist, deprovisioning command)

- An alternative implementation we discussed is uploading the CSV from the RA. Only SRAA\'s would be allowed to do this. The advantage is that an authorised and authenticator actor would be available (the SRAA) and that a migration would no require console access. However more work to implement. Long running batch processes from HTTP client is much more work to implement than a console command. Decided not to do this now, can be added later reusing most of the implementation.

- Logging: log at least the start and end of the migration of a single identity at a sufficiently high level that it would be recorded in the normal logging. Also log each migrated token.

Note: event naming is still open.

Next steps: split this "epic" in separate stories ( @michielkodde ). Refine these stories before starting the implementation ( @phavekes  @pmeulen @michielkodde ).
We estimate 5 days of developer work for stories, refinement and implementation. (Pieter van der Meulen - Apr 22, 2021)
